rust-runtime/aws-smithy-http-server-python/src/util/collection.rs

```diff
@@ -0,0 +1,339 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//! Provides Rust equivalents of [collections.abc] Python classes.
+//!
+//! Creating a custom container is achived in Python via extending a `collections.abc.*` class:
+//! ```python
+//! class MySeq(collections.abc.Sequence):
+//!     def __getitem__(self, index):  ...  # Required abstract method
+//!     def __len__(self):  ...             # Required abstract method
+//! ```
+//! You just need to implement required abstract methods and you get
+//! extra mixin methods for free.
+//!
+//! Ideally we also want to just extend abstract base classes from Python but
+//! it is not supported yet: <https://github.com/PyO3/pyo3/issues/991>.
+//!
+//! Until then, we are providing traits with the required methods and, macros that
+//! takes those types that implement those traits and provides mixin methods for them.
+//!
+//! [collections.abc]: https://docs.python.org/3/library/collections.abc.html
+
+use pyo3::PyResult;
+
+/// Rust version of [collections.abc.MutableMapping].
+///
+/// [collections.abc.MutableMapping]: https://docs.python.org/3/library/collections.abc.html#collections.abc.MutableMapping
+pub trait PyMutableMapping {
+    type Key;
+    type Value;
+
+    fn len(&self) -> PyResult<usize>;
+    fn contains(&self, key: Self::Key) -> PyResult<bool>;
+    fn keys(&self) -> PyResult<Vec<Self::Key>>;
+    fn values(&self) -> PyResult<Vec<Self::Value>>;
```

Comment:
I'm wondering if this is a footgun.
In Python you'd expect iterating over an existing collection to be "free", while our implementation triggers a deep clone of the entire map.

Reply:
I.e. as a Python developer, I wouldn't expect this:
```python
for elem in map:
    assert elem in elems
```
to trigger a deep clone of all elements in the map.

Reply:
I think this is a good first step, but me and Burak already talked about understanding what is needed to prevent the deep clone and borrow the map so that it becomes cheaper. I am not even sure if this is possible with the current PyO3 support.

Reply:
Btw, if you are already tracking this issue feel free to merge and we can revisit it later. But perhaps worth calling it out in the docs for those methods.

Reply:
👍 I've added comments to the related methods: https://github.com/awslabs/smithy-rs/pull/1871/commits/c8bd36852b3be76914b5637f855a1e9e7a972571.


> This would be less of an issue If both the header names and their values happened to be behind smart pointers.

We are converting them to `String`s while crossing to Python: https://github.com/awslabs/smithy-rs/blob/c8bd36852b3be76914b5637f855a1e9e7a972571/rust-runtime/aws-smithy-http-server-python/src/middleware/header_map.rs#L56-L66